### PR TITLE
mignotte bound applied in `dup_zz_zassenhaus(f, K)`

### LINKED ISSUE #19253: sympy.polys.factortools.dmp_zz_mignotte_bound improvement
The method `dup_zz_mignotte_bound(f, K)` can be significantly improved by using the **Knuth-Cohen bound** instead. After our research with Prof. Ag.Akritas we have implemented the Knuth-Cohen bound among others, and compare them among dozens of polynomials with different degree, density and coefficients range. Considering the results and the feedback from Mr.Kalevi Suominen, our proposal is that the mignotte_bound should be replaced by the knuth-cohen bound.
Also, `dmp_zz_mignotte_bound(f, u, K)` for mutli-variants polynomials should be replaced appropriately.

### PR DESCRIPTION
**Pull request by lagamura**
#### References to other Issues or PRs
Application of PR 19254

Fixes #19253 partially.

#### Brief description of what is fixed or changed
function `dup_zz_zassenhaus(f, K)` instead of calling `dup_zz_maignotte_bound(f, K)` calculates locally the old bound : `(abs(K.sqrt(K(n + 1))*2**n*A*b).`

#### Other comments
By changing the aforementioned line, there is an improvement on `factor(f)` to 10% on time. 

#### Release Notes

* polys
   * improvement of `dup_zz_zassenhaus(f, K)`

### CODE REVIEW

**Comment by lagamura** [comment]
@jksuom give a look on this

File: sympy/polys/factortools.py [hunk 1]
```diff
@@ -323,7 +323,7 @@ def dup_zz_zassenhaus(f, K):
     fc = f[-1]
     A = dup_max_norm(f, K)
     b = dup_LC(f, K)
-    B = int(abs(K.sqrt(K(n + 1))*2**n*A*b))
+    B = int(dup_zz_mignotte_bound(f, K))
```
Comment (jksuom) [inline sympy/polys/factortools.py:326]: It looks like the improved Mignotte bound cannot be used here. There is a test https://github.com/sympy/sympy/blob/0515aeec9ff25f5abb831f0d54a6eb6704084551/sympy/polys/factortools.py#L402
that fails if `B` is too small. I haven't seen the original documentation but it seems that this `B` should not be changed.

File: sympy/polys/factortools.py [hunk 2]
```diff
@@ -323,7 +323,11 @@ def dup_zz_zassenhaus(f, K):
     fc = f[-1]
     A = dup_max_norm(f, K)
     b = dup_LC(f, K)
-    B = 2*int(abs(dup_zz_mignotte_bound(f, K)))
+    if Poly(f).is_univariate:
```
Comment (jksuom) [inline sympy/polys/factortools.py:326]: There is no need for this test. This method is for univariate polynomials.

**Comment by lagamura** [comment]
Can someone help with that log: (link)[https://travis-ci.org/github/sympy/sympy/jobs/686930526]
Type error seems strange as we have not change anything on types of the function.

**Comment by lagamura** [comment]
It seems we are failing in some specific examples on travic ci. Although when we are running them locally we are passing them successfully as you can see by the attached file. Is there any possibility that travic ci has a bug?
[test2_latest.zip](https://github.com/sympy/sympy/files/4627632/test2_latest.zip)

**Comment by lagamura** [comment]
We made some changes, and now Travis CI gives timeout because no output was recieved.
This means something like an infinite loop? How can i debug this?

**Comment by jksuom** [comment]
It looks like that is happening in polytools (and maybe elsewhere). I would run `bin/test -v polytools` in order to see which test is hanging. Then I would run that test in a debugger and  interrupt it after a while. The stack trace will usually show periodicity that will help to locate the looping lines. I this case, those will probably be the changed lines so that putting a breakpoint there could also help.

**Comment by smichr** [comment]
Just read an [article that might be of interest](https://www.quantamagazine.org/new-math-measures-the-repulsive-force-within-polynomials-20200514/) in which in says,

>Dimitrov proved that all non-cyclotomic polynomials must have a root that’s at least log 2 divided by four times the degree of the polynomial outside the unit circle; that is, (log 2)/(4d)

File: sympy/polys/factortools.py [hunk 3]
```diff
@@ -169,7 +169,7 @@ def dup_zz_mignotte_bound(f, K):
     delta2 = _ceil(delta / 2)
 
     # euclidean-norm
-    eucl_norm = K.sqrt( sum( [cf**2 for cf in f] ) )
+    eucl_norm = _sqrt( sum( [cf**2 for cf in f] ) )
```
Comment (jksuom) [inline sympy/polys/factortools.py:172]: This will change the type of `eucl_norm` that is an element of `K`. What is the purpose of the change?

Comment (lagamura) [reply sympy/polys/factortools.py:172]: Alkis asked me to change this, because he founds that now it is passing an test. Although for me it stacks again in an infinite loop on `test_Poly_lift` file. Now i will try to find the periodicity you mentioned above

**Comment by czgdp1807** [comment]
@lagamura Are you still working on this? Thanks for your contributions. Please do not close this PR.

**Comment by jksuom** [comment]
I think that this can be closed. It was based on a bound that did not work in this connection.

**Comment by jksuom** [comment]
Closing.
